### PR TITLE
Beta

### PR DESCRIPTION
**Pull request by ProjectBarks**
# Completed
- Uploader 
- Tile progress 
- View button 
- Token creation

# TODO
- Bug fixes
- Make custom options work
- Add terrain options back in 

# How to install?
- [Install WebODM through docker](https://github.com/OpenDroneMap/WebODM#getting-started)
- Clone the repo into the `plugins` directory
- Start the project using `./webodm start --dev` 
- Download a task from [WebODM Demo](https://demo.webodm.org/dashboard/) (using all)
- Import zip into WebODM instance

# Screenshots 
image
image
image

### CODE REVIEW

**Comment by mramato** [comment]
Awesome, will check this out later this morning.

**Comment by mramato** [comment]
> * Download a task from [WebODM Demo](https://demo.webodm.org/dashboard/) (using all)

This just brings me to a login page, is there a demo account or something they advertise?  I couldn't find anything.

I had some local photogrammetry I processed, but I still don't see the ion button?  I believe you left out details about configuring ion?

**Comment by mramato** [comment]
You have instructions missing for setting up the key, and it wasn't clear to me what I had to do (until I noticed the Cesium ion tab on the left side of the screen)

**Comment by mramato** [comment]
1. Once I configured a token, the ion dropbutton showed up on existing tasks.
2. When a new task is finished, the ion dropbutton is missing until I refresh the page?
3. I tried to upload an orthophoto but it seems to have silently fail (it shows `Uploading` as a status but it doesn't look anything is actually happening).  Any ideas?

**Comment by ProjectBarks** [comment]
Yeah my bad I was throwing those instructions together pretty late

**Comment by ProjectBarks** [comment]
# Required
- [ ] Finish passing options to uploads
- [ ] Reduce polling for error stares
- [ ] Add error information expanding panel in task list
- [ ] Finish documentation (README, Building Instructions, Structure)

**Comment by mramato** [comment]
At a high level, that's a good start.  Some other things to think about

• Do we validate the token at all?
• General error handling/polish, i.e. What happens if I change a token while a job is running? Should that even be allowed?
• Can we show the ion button disabled with a link/text for instructions to install the access token?
• Do we really need the separate tasks UX if "View in ion" shows me the progress anyway?  Similarly, do we need to wait until the task is finished before "View in ion" shows up? (I guess it depends on how error handling will work

In general, I would concentrate on the core features/functionality and get everything working in ideal cases and then we can go through errors and edge cases and see what's left.

**Comment by ProjectBarks** [comment]
> Do we validate the token at all?
No but there is no way to do that through the API 

> General error handling/polish, i.e. What happens if I change a token while a job is running? Should that even be allowed?
Tokens are serialized into tasks which are sent to workers so as long as the token was valid before submissions the task should run fine.

>Can we show the ion button disabled with a link/text for instructions to install the access token?
Sure 

> Do we really need the separate tasks UX if "View in ion" shows me the progress anyway? Similarly, do we need to wait until the task is finished before "View in ion" shows up? (I guess it depends on how error handling will work
Yeah I think separation makes sense since clicking a button should not have widely different outcomes. But I agree the UX there is not as polished as it could be.

**Comment by mramato** [comment]
> No but there is no way to do that through the API

This is a fair point, as long as we properly handle the case where the upload fails because of auth and inform the user to check their token.

**Comment by ProjectBarks** [comment]
@mramato this version is ready for feedback

**Comment by mramato** [comment]
1. How are we keeping track of what has been loaded in ion?  I uploaded a few assets but now it doesn't think any have been uploaded.  Does setting the token reset anything?
2. If I set an incorrect token, the page just crashes when pick something in the ion dropdown
3. Does the Surface Model default to Ellipsoid because we know that's what ODM creates, or should it default to Mean Sea Level (as ion defaults to?)

**Comment by mramato** [comment]
Importing a task does not cause the Cesium ion dropdown to appear until you refresh the page.

**Comment by mramato** [comment]
I just uploaded a terrain model and realized I screwed up and didn't choose the right setting, but since we aren't letting people upload the same model twice right now, there's no way for me to re-do it.  Any thoughts on how to address this?

**Comment by ProjectBarks** [comment]
> How are we keeping track of what has been loaded in ion? I uploaded a few assets but now it doesn't think any have been uploaded. Does setting the token reset anything?

Each task has metadata associated with what is uploaded. Upon load we do a sanity check to see what does not currently exist via the ion API. The only time this is updated is on reload as its pretty heavy to be constantly checking for if something has been deleted.

> If I set an incorrect token, the page just crashes when pick something in the ion dropdown

That should be added as a handled case

> Does the Surface Model default to Ellipsoid because we know that's what ODM creates, or should it default to Mean Sea Level (as ion defaults to?)

We have no reasoning for defaulting. I can change that if it makes sense 

> Importing a task does not cause the Cesium ion dropdown to appear until you refresh the page.

I dont have a good solution for that, we can figure something out.

> I just uploaded a terrain model and realized I screwed up and didn't choose the right setting, but since we aren't letting people upload the same model twice right now, there's no way for me to re-do it. Any thoughts on how to address this?

Yes you can as long as you delete the model on ion. There is no reason to reupload, maybe there is a good idea for how to handle this?

**Comment by mramato** [comment]
> We have no reasoning for defaulting. I can change that if it makes sense

The reason is that most data is `Mean Sea Level`, it could be that in ODM all data is Mean Sea Level (in which case there's no reason to even present it as an option.  Can we reach out to them or check their doc to confirm?



> Yes you can as long as you delete the model on ion. There is no reason to reupload, maybe there is a good idea for how to handle this?

Okay, cool.  So deleting on ion will cause the next page refresh to not show the asset.  Sounds good.

**Comment by ProjectBarks** [comment]
> The reason is that most data is Mean Sea Level, it could be that in ODM all data is Mean Sea Level (in which case there's no reason to even present it as an option. Can we reach out to them or check their doc to confirm?

Totally misunderstood that. Ill check that now

**Comment by mramato** [comment]
> Each task has metadata associated with what is uploaded. Upon load we do a sanity check to see what does not currently exist via the ion API. The only time this is updated is on reload as its pretty heavy to be constantly checking for if something has been deleted.

Okay, I probably deleted the assets in my ion account and didn't realize it would be reflected here (which sounds like the preferred way to have it work).

**Comment by mramato** [comment]
In addition to the issue I showed you offline, we probably want to remove the watermask checkbox as well. I don't see a use case where anyone would want to enable it via ODM. (We can always add it back later).

**Comment by mramato** [comment]
@ProjectBarks is this ready for another look?  Please remember to leave a comment when a PR is ready for another round of review.

**Comment by ProjectBarks** [comment]
@mramato yes

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Code of Conduct
+
+One of Cesium's strengths is our community. Our contributors and users are pushing the 3D geospatial field to amazing new levels. We rely on an open, friendly, inclusive environment to facilitate this. As such, we follow the [Contributor Covenant](http://contributor-covenant.org/)'s [Code of Conduct](http://contributor-covenant.org/version/1/4/code-of-conduct.md) to ensure a harassment-free experience in the Cesium community. Any unacceptable behavior can be confidentially sent to the core team at pcozzi@agi.com.
```
Comment (mramato) [inline CODE_OF_CONDUCT.md]: This was changed as part of the Blender add-on review process, please update to match the next in https://raw.githubusercontent.com/AnalyticalGraphicsInc/cesium-ion-blender-addon/master/CODE_OF_CONDUCT.md

**Comment by mramato** [comment]
Looks like both LICENSE and README were accidentally changed to include the executable bit.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,9 @@
-# cesium-ion-webodm-exporter
-WebODM plugin for uploading and tiling data with Cesium ion. https://cesium.com
+<p align="center">
+  <img src="https://github.com/AnalyticalGraphicsInc/cesium/wiki/logos/Cesium_Logo_Color.jpg" width="50%" />
+</p>
+
+The Cesium ion WebODM add-on enables you to publish and stream even the most massive of 3D Models on the web with 3D Tiles.
```
Comment (mramato) [inline README.md]: I know the Blender readme was used for a base here, which is fine, but please update the copy to include more ODM specific content and use their teminology (plugin not add-in, etc..)  Also ODM does more than just 3D Models, so we need to include everything this plug-in does.

File: ThrirdParty.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
+[
```
Comment (mramato) [inline ThrirdParty.json]: You misspelled the file name.

Please include all direct, non-dev dependencies from package.json

File: documentation/developer-guide/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
+# Development
+
+Follow these steps to run the addon directly from source so that your changes will be reflected live.
+
+## Installation
+
+1. Clone the WebODM repository `git clone git@github.com:OpenDroneMap/WebODM.git`
+
+1. Clone the plugin repository into the `plugins` folder of WebODM.
```
Comment (mramato) [inline documentation/developer-guide/README.md]: I think changing `the plugin` -> `this` is more clear.

File: README.md (hunk 2 above)
Comment (ProjectBarks) [reply README.md]: Ill fix this in a follow up PR

File: documentation/developer-guide/README.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
+# Development
```
Comment (mramato) [inline documentation/developer-guide/README.md]: Please rename the file paths so they are consistent across Cesium repositories.

File: documentation/developer-guide/README.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    1. Ensure that docker is installed on your system; if not [follow this guide](https://docs.docker.com/v17.12/install/).
+
+    1. In the root of WebODM run
+        - **OSX - Linux** - `./webodm.sh start --dev`
+        - **Windows** - No support, checkout [documentation](https://github.com/OpenDroneMap/WebODM#getting-started).
+
+## Setup
+
+1. Grab a token with, `assets:list, assets:read, assets:write, geocode` permissions from [Cesium ion](https://cesium.com/ion/tokens).
```
Comment (mramato) [inline documentation/developer-guide/README.md]: There shouldn't be any reason to have `geocode` permissions, right?

File: manifest.json [hunk 7]
```diff
@@ -0,0 +1,13 @@
+{
+	"name": "Cesium ion",
+	"webodmMinVersion": "0.6.0",
+	"description": "Upload and tile models with Cesium ion.",
```
Comment (mramato) [inline manifest.json]: This plugin does more than just models, does ODM have a generic term for what it creates?

File: manifest.json [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+	"version": "1.0.0",
+	"author": "Analytical Graphics, Inc.",
```
Comment (mramato) [inline manifest.json]: We can now change this to Cesium GS, Inc

File: public/defaults/AssetStyles.jsx [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+		name: "Orthophoto",
+		icon: "fa-map-o"
+	},
+	[AssetType.TERRAIN_MODEL]: {
+		name: "Terrain Model",
+		icon: "fa-area-chart"
+	},
+	[AssetType.SURFACE_MODEL]: {
+		name: "Surface Model",
+		icon: "fa-table"
```
Comment (mramato) [inline public/defaults/AssetStyles.jsx]: Shouldn't this use `fa-area-chart` too?

File: public/package.json [hunk 10]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "WebODM",
```
Comment (mramato) [inline public/package.json]: Please update all of these fields

File: templates/load_buttons.js [hunk 11]
```diff
@@ -0,0 +1,12 @@
+PluginsAPI.Dashboard.addTaskActionButton(
+	["${app_name}/build/TaskView.js"],
+	function(args, TaskView) {
+		console.log(args);
```
Comment (mramato) [inline templates/load_buttons.js]: Is this log intentional or left over from debugging?

**Comment by mramato** [comment]
I didn't look heavily at the code, but I think that's all I have.  You mentioned a separate PR to clean things up, feel free to branch off this and PR into it and then we'll merge the whole thing up to master.

File: public/defaults/AssetStyles.jsx (hunk 9 above)
Comment (ProjectBarks) [reply public/defaults/AssetStyles.jsx]: WebODM uses both icons for Surface Model but yeah I would agree

File: public/package.json (hunk 10 above)
Comment (ProjectBarks) [reply public/package.json]: I removed this and only store scripts and dev dependencies in favor a root manifest only

File: templates/load_buttons.js (hunk 11 above)
Comment (ProjectBarks) [reply templates/load_buttons.js]: EH 50/50. I use it so much when debugging that I always put it back in but yeah Ill remove it

**Comment by mramato** [comment]
Going to merge this since functionality seems to be in pretty good shape and we can make further tweaks in new PRs.  Thanks @ProjectBarks!

**Pull request merged**